libpysal/weights/raster.py

```diff
@@ -0,0 +1,297 @@
+
+__all__ = ['da_checker', 'da2W', 'da2WSP', 'w2da', 'da2W2', 'da2WSP2']
+
+
+def da2W(da, criterion, **kwargs):
+    """
+    Create a W object from rasters(xarray.DataArray)
+    Parameters
+    ----------
+    da         : xarray.DataArray
+                 raster file accessed using xarray.open_rasterio method
+    criterion  : {"rook", "queen"}
+                 option for which kind of contiguity to build
+    Returns
+    -------
+    w    : libpysal.weights.W
+           instance of spatial weights class W
+    """
+    if criterion != 'rook':
+        rook = False
+    else:
+        rook = True
+    w = lat2W(*da[0].shape, rook, **kwargs)
+    ser = da.to_series()
+    id_order = np.where(ser != da.nodatavals[0])[0]
+    w = w_subset(w, id_order)
+    ser = ser[ser != da.nodatavals[0]]
+    w.coords = ser.index
+    attrs = da.attrs
+    attrs["dims"] = da.dims
+    attrs["shape"] = da.shape
+    w.attrs = attrs
+    return w
+
+
+def da2W2(da, criterion, **kwargs):
+    """
+    Create a W object from rasters(xarray.DataArray)
+
+    Parameters
+    ----------
+    da         : xarray.DataArray
+                 raster file accessed using xarray.open_rasterio method
+    criterion  : {"rook", "queen"}
+                 option for which kind of contiguity to build
+    Returns
+    -------
+    w    : libpysal.weights.W
+           instance of spatial weights class W
+    """
+    ser = da.to_series()
+    indices = np.where(da.data.flatten() == da.nodatavals[0])[0]
+    id_order = np.where(da.data.flatten() != da.nodatavals[0])[0]
+    sw = _raster2SW(*da[0].shape, indices, criterion)
+    w = WSP(sw, id_order.tolist()).to_W(**kwargs)
+    ser = ser[ser != da.nodatavals[0]]
+    w.coords = ser.index
+    attrs = da.attrs
+    attrs["dims"] = da.dims
+    attrs["shape"] = da.shape
+    w.attrs = attrs
+    return w
+
+
+def da2WSP(da, criterion, **kwargs):
+    """
+    Generate a WSP object from rasters(xarray.DataArray)
+    Parameters
+    ----------
+    da         : xarray.DataArray
+                 raster file accessed using xarray.open_rasterio method
+    criterion  : {"rook", "queen"}
+                 option for which kind of contiguity to build
+    Returns
+    -------
+    sw    : libpysal.weights.WSP
+           instance of spatial weights class WSP
+    """
+    sw = lat2SW(*da[0].shape, criterion)
+    ser = da.to_series()
+    id_order = np.where(ser != da.nodatavals[0])[0]
+    indices = np.where(ser == da.nodatavals[0])[0]
+    mask = np.ones((sw.shape[0],), dtype=np.bool)
+    mask[indices] = False
+    for _ in range(2):
+        sw = sw[mask]
+        sw = sw.transpose()
+        sw = sw.tocsr()
+    sw = WSP(sw)
+    sw.id_order = id_order
+    ser = ser[ser != da.nodatavals[0]]
+    sw.coords = ser.index
+    attrs = da.attrs
+    attrs["dims"] = da.dims
+    attrs["shape"] = da.shape
+    sw.attrs = attrs
+    return sw
+
+
+def da2WSP2(da, criterion):
```

Comment:
I was able to optimize `lat2SW` using `numba` though since scipy is not supported therefore the performance gain is not huge. this function uses the new `raster2WS`